Fix item duplication on belts

danth opened the pull request:
When a belt points directly into another belt, the last item to cross the boundary may be duplicated when the chunk is unloaded. I believe this happens because the ending doesn't mark the inventory as changed after removing the item.

This is my first time contributing to a Minecraft mod, so please double check that I'm correct before merging.

Fixes #7242
Fixes #4526

VoidLeech commented:
The language in this PR doesn't sound too convincing, although it sounds like a plausible explanation. Have you (somewhat consistently) reproduced the bug without the PR where you couldn't with?
Code change wise I'd guess this would also fix #4526, can you check?

danth commented:
Yes, I can reproduce the bug as follows:

1. In a flat world, away from the spawn chunks, set up a few long belts feeding into each other
2. Drop an item onto the first belt, follow it to the other end and pick it up
3. Teleport far away, wait a few seconds
4. Teleport back and the item reappears near the end of some of the belts

I did this experiment about 5-10 times, duplication happened all but once before this change and never happened after.

I also had print statements at various places in `BeltInventory`, these showed the item being removed from each inventory as it moved between the belts, but the call to `read` after returning to the chunk had the item in the inventory again.

#4526 does sound like the same issue, since going to the nether is another way to unload chunks - assuming that the `INSERT` ending also handles funnels as well as other belts. I'll test this properly soon.

danth commented:
I can reproduce #4526 only when the target inventory is in a different chunk, like this:

![Belt and funnel in one chunk, barrel in adjacent chunk](https://github.com/user-attachments/assets/5adf06c9-ab3d-4029-a438-865ea0b858fd)

This PR doesn't fix that issue, but I have a feeling its cause is similar. When the barrel is in the same chunk, the barrel presumably calls `setChanged` after receiving the item, and causes the entire chunk to be saved. When the barrel is in a different chunk, the chunk with the belt is not saved and so its inventory rolls back after teleporting away.

The patch below fixes it with a single item on the belt, but I have a feeling it doesn't cover all possible cases, such as the funnel only taking part of a stack.

```patch
diff --git a/src/main/java/com/simibubi/create/content/kinetics/belt/transport/BeltInventory.java b/src/main/java/com/simibubi/create/content/kinetics/belt/transport/BeltInventory.java
index d1a870b70..964e861e6 100644
--- a/src/main/java/com/simibubi/create/content/kinetics/belt/transport/BeltInventory.java
+++ b/src/main/java/com/simibubi/create/content/kinetics/belt/transport/BeltInventory.java
@@ -110,6 +110,7 @@ public class BeltInventory {
 			if (currentItem.stack.isEmpty()) {
 				iterator.remove();
 				currentItem = null;
+				belt.setChanged();
 				continue;
 			}
 ```

danth committed:
Fix item duplication on belts

Belts were not always saved after their inventory changed, leading to
duplication of items when the chunk was reloaded.

danth commented:
Okay, this PR now fixes #4526 too, including when the funnel takes part of a stack.

I also fixed a similar issue with tunnels:

![2025-04-28_11 28 04](https://github.com/user-attachments/assets/077ab7fb-59f4-4a8f-aa2b-dc8cf0964460)

The inventory of the main belt wasn't saved after the stack passed through the tunnel, but the item which split off was, so the stack would pass through again after reloading the chunk.

simibubi commented:
Thank you!
We might have to look out for side-effects in belt performance. If the additional comparator updates these changes create become noticeable, we should pivot to sendData and some form of markChunkDirty
